great_expectations/expectations/metrics/util.py

```diff
@@ -1058,3 +1059,83 @@ def get_sqlalchemy_source_table_and_schema(
+    """
+    Builds unexpected_index list for Pandas Dataframe in situation where the named
+    columns is also a named index. This method handles the case when there are multiple named indices.
+    Args:
+        df: reference to Pandas dataframe
+        expectation_domain_column_name: column that Expectation is being run for
+        unexpected_index_column_names: column_names for indices, either named index or unexpected_index_columns
+
+    Returns:
+        List of Dicts that contain ID/PK values
+    """
+    if not expectation_domain_column_name:
+        raise ge_exceptions.MetricResolutionError(
+            message=f"Error: The domain column is currently set to None. Please check your configuration.",
+            failed_metrics=["unexpected_index_list"],
+        )
+    index_names: List[str] = df.index.names
+    unexpected_index_list: List[Dict[str, Any]] = list()
+    unexpected_indices: List[Union[int, str]] = list(df.index)
+    for index in unexpected_indices:
+        primary_key_dict: Dict[str, Any] = dict()
+        # domain column first
+        primary_key_dict[expectation_domain_column_name] = df.at[
+            index, expectation_domain_column_name
+        ]
+        for column_name in unexpected_index_column_names:
+            if column_name not in index_names:
+                raise ge_exceptions.MetricResolutionError(
+                    message=f"Error: The column {column_name} does not exist in the named indices. Please check your configuration",
+                    failed_metrics=["unexpected_index_list"],
+                )
+            tuple_index = index_names.index(column_name, 0)
+            primary_key_dict[column_name] = index[tuple_index]
+        unexpected_index_list.append(primary_key_dict)
+    return unexpected_index_list
+
+
+def get_unexpected_indices_for_single_pandas_named_index(
+    df: pd.DataFrame,
+    expectation_domain_column_name: str,
+    unexpected_index_column_names: List[str],
+) -> List[Dict[str, Any]]:
+    """
+    Builds unexpected_index list for Pandas Dataframe in situation where the named
+    columns is also a named index. This method handles the case when there is a single named index.
+    Args:
+        df: reference to Pandas dataframe
+        expectation_domain_column_name: column that Expectation is being run on.
+        unexpected_index_column_names: column_names for indices, either named index or unexpected_index_columns
+
+    Returns:
+        List of Dicts that contain ID/PK values
+
+    """
+    unexpected_index_values_by_named_index: List[Union[int, str]] = list(df.index)
+    unexpected_index_list: List[Dict[str, Any]] = list()
+
+    for index in unexpected_index_values_by_named_index:
+        primary_key_dict: Dict[str, Any] = dict()
+        # domain column first
+        primary_key_dict[expectation_domain_column_name] = df.at[
+            index, expectation_domain_column_name
+        ]
+        for column_name in unexpected_index_column_names:
```

Comment:
Is `unexpected_index_column_names` always of length 1 in this function? If so, I'd assert that at the beginning of this function. Then you can remove this inner `for` loop.

Reply:
Yes, and we would want to assert this. Great idea. thank you

Reply:
Added assertion